contracts/zero-ex/contracts/src/features/MultiplexFeature.sol

```diff
@@ -0,0 +1,799 @@
+                batchFillData.outputToken = IERC20TokenV06(fillData.tokens[i + 1]);
+                uint256 ethValue;
+                (batchFillData.calls, ethValue) = abi.decode(
+                    wrappedCall.data,
+                    (WrappedBatchCall[], uint256)
+                );
+                // Do not spend more than the remaining ETH.
+                ethValue = LibSafeMathV06.min256(ethValue, remainingEth);
+                if (IERC20TokenV06(fillData.tokens[i]).isTokenETH()) {
+                    batchFillData.sellAmount = LibSafeMathV06.min256(
+                        outputTokenAmount,
+                        ethValue
+                    );
+                } else {
+                    batchFillData.sellAmount = outputTokenAmount;
+                }
+                // Decrement the remaining ETH.
+                remainingEth -= ethValue;
+                uint256 leftoverEth;
+                (outputTokenAmount, leftoverEth) = _batchFill(batchFillData, ethValue);
+                // Add back any ETH that was unused by the `_batchFill`.
+                remainingEth += leftoverEth;
+                nextTarget = address(0);
+            } else if (wrappedCall.selector == IEtherTokenV06.deposit.selector) {
+                uint256 ethValue = LibSafeMathV06.min256(outputTokenAmount, remainingEth);
+                // Wrap ETH.
+                weth.deposit{value: ethValue}();
+                nextTarget = _computeHopRecipient(fillData.calls, i);
+                weth.transfer(nextTarget, ethValue);
+                remainingEth -= ethValue;
+            } else if (wrappedCall.selector == IEtherTokenV06.withdraw.selector) {
+                // Unwrap WETH and send to `msg.sender`.
+                weth.withdraw(outputTokenAmount);
+                msg.sender.transfer(outputTokenAmount);
+                nextTarget = address(0);
+            } else {
+                revert("MultiplexFeature::_multiHopFill/UNRECOGNIZED_SELECTOR");
+            }
+        }
+    }
+
+    // Similar to the UniswapFeature, but with a couple of differences:
+    // - Does not perform the transfer in if `pairAddress` is given,
+    //   which indicates that the transfer in was already performed
+    //   in the previous hop of a multi-hop fill.
+    // - Does not include a minBuyAmount check (which is performed in
+    //   either `batchFill` or `multiHopFill`).
+    // - Takes a `recipient` address parameter, so the output of the
+    //   final `swap` call can be sent to an address other than `msg.sender`.
+    function _sellToUniswap(
+        address[] memory tokens,
+        uint256 sellAmount,
+        bool isSushi,
+        address pairAddress,
+        address recipient
+    )
+        public
+        returns (uint256 outputTokenAmount)
+    {
+        require(tokens.length > 1, "MultiplexFeature::_sellToUniswap/InvalidTokensLength");
+
+        if (pairAddress == address(0)) {
```

Comment:
Are there cases when `pairAddress != address(0)`?

Reply:
yes, in `_multiHopFill` the pair address for the next hop may be pre-computed and stored in the `nextTarget` variable. This is for a scenario like a PLP => Uniswap multihop fill, where we want the PLP to transfer the bought tokens directly into the Uniswap pair contract

Reply:
Ah, I see. Should we perform any validation ~ is it an issue if the tokens are transferred somewhere other than a Uniswap pool?

Reply:
that should never happen, since `nextTarget` would have been set to the Uniswap pair address as computed using `create2` in the previous hop. At worst the pair contract doesn't exist, but it can't go anywhere unexpected